File: arrays/Cyclically_Rotating_A_Grid.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 Beats 100%
 */
var rotateGrid = function(grid, k) {
    let layer = Math.min(grid.length / 2, grid[0].length / 2);
    const m = grid.length;
    const n = grid[0].length;
    layer = Math.min(m, n) / 2; //As evident from the example images, the number of layers depends on the min. of col_length and row_length

    //Each layer gets processed independently.
    for (let i = 0; i < layer; i++) {

        let firstRow = grid[i].slice(i, n - i);
        const rowLength = firstRow.length;

        let lastElementsMiddleRows = [];
        for (let j = i + 1; j < m - i - 1; j++) {
            lastElementsMiddleRows.push(grid[j][n - i - 1]);
        }
        const colLength = lastElementsMiddleRows.length;

        let lastRowReversed = grid[m - i - 1].slice(i, n - i).reverse();

        let firstElementsMiddleRowsReversed = [];
        for (let j = m - i - 2; j > i; j--) {
            firstElementsMiddleRowsReversed.push(grid[j][i]);
        }

        let combinedList = firstRow.concat(
            lastElementsMiddleRows,
            lastRowReversed,
            firstElementsMiddleRowsReversed
        );

        const shiftReq = k % combinedList.length; //This is very important, don't just shift it by k. If k>list length, it'll throw error
        combinedList = combinedList.slice(shiftReq).concat(combinedList.slice(0, shiftReq));

        // After shifting, the elements need to be reassigned their positions
        grid[i].splice(i, rowLength, ...combinedList.slice(0, rowLength)); // First row

        for (let j = i + 1; j < m - i - 1; j++) {
            grid[j][n - i - 1] = combinedList[rowLength + j - i - 1]; // Last elements of middle rows
        }

        grid[m - i - 1].splice(i, rowLength, ...combinedList.slice(rowLength + colLength, rowLength + colLength + rowLength).reverse()); // Last row

        for (let j = m - i - 2; j > i; j--) {
            grid[j][i] = combinedList[rowLength + colLength + rowLength + m - i - 2 - j]; // First elements of middle rows
        }
    }

    return grid;
};

File: arrays/Cyclically_Rotating_A_Grid.py
'''
https://leetcode.com/problems/cyclically-rotating-a-grid/description/
You are given an m x n integer matrix grid​​​, where m and n are both even integers, and an integer k.

The matrix is composed of several layers, which is shown in the below image, where each color is its own layer:

Input: grid = [[1,2,3,4],[5,6,7,8],[9,10,11,12],[13,14,15,16]], k = 2
Output: [[3,4,8,12],[2,11,10,16],[1,7,6,15],[5,9,13,14]]
Explanation: The figures above represent the grid at every state.

Constraints:

m == grid.length
n == grid[i].length
2 <= m, n <= 50
Both m and n are even integers.
1 <= grid[i][j] <= 5000
1 <= k <= 109

#This is one of the best solutions I have done. It beats 77% of the users. They key is to always take it simple, and break it into simple chunks. If you're thinking a complicated way, then 
most likely a simpler solution exists.
'''

class Solution:
    def rotateGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        layer = min( len(grid)/2, len(grid[0])/2 )

        m,n = len(grid), len(grid[0])
        layer = min(m, n)//2

        for i in range(0,layer): #Each layer get cycled independently. 
          #A grid consists of, first row, the last elements of middle row, the last row reversed and the first elements of the middle rows reversed.
            first_row = grid[i][i:n-i]
            row_length = len(first_row)
            last_elements_middle_rows = [row[n-i-1] for row in grid[i+1:m-i-1]]
            col_length = len(last_elements_middle_rows)
            last_row_reversed = grid[m-i-1][i:n-i][::-1]
            first_elements_middle_rows_reversed = [row[i] for row in grid[m-i-2:i:-1]]
            combined_list = first_row + last_elements_middle_rows + last_row_reversed + first_elements_middle_rows_reversed #This is the ith layer of the grid, starting from outermost layer
            shift_req = k%len(combined_list) #Because it k>len(combined_list), the shift function won't work.
            combined_list = combined_list[shift_req:]+combined_list[:shift_req] #We shift the elements k places to the left.

            first_row = combined_list[:row_length] #After shifting, the elements need to be reassigned their positions
            last_elements_middle_rows = combined_list[row_length:row_length+col_length]
            last_row_reversed = combined_list[row_length+col_length:row_length+col_length+row_length]
            first_elements_middle_rows_reversed = combined_list[row_length+col_length+row_length:]

            grid[i][i:n-i] = first_row #Working alright (First row)

            for j in range(i+1,m-i-1): ##Last elements of middle rows assignment
                grid[j][n-i-1] = last_elements_middle_rows[j-i-1] #Working alright

            grid[m-i-1][i:n-i] = last_row_reversed[::-1] #Working alright #The last row (reversed) assignment, that's why last_row_reversed is reversed using [::-1]
            
            for j in range(m-i-2,i,-1): #First elements of middle rows
                grid[j][i] = first_elements_middle_rows_reversed[m-i-2-j]

        return grid
